### ISSUE TITLE
Using autograd for matrix factorization

### ISSUE THREAD

**Issue by nipunbatra** [HIPS/autograd#207]
Hi,
I was trying to use autograd for a simple matrix factorization. `A (MXN) -> W (MXk). H (KXN)`. Here is the code that I have written. The cost is the frobenius norm of the difference between A and W.H. When I compute autograd on this cost, autograd returns a matrix corresponding to the shape of W. How can we get the gradient value for both W and H?

```python
import autograd.numpy as np
import pandas as pd

A = np.array([[3, 4, 5, 2],
                   [4, 4, 3, 3],
                   [5, 5, 4, 4]], dtype=np.float32)

shape = A.shape #M, N
rank = 2 #k

def cost(W, H):
    pred = np.dot(W, H)
    return np.sqrt(((pred - V).flatten() ** 2).mean(axis=None))

H =  np.random.randn(rank, shape[1])
W =  np.random.randn(shape[0], rank)

# Testing cost
# cost(W, H) 
# Returned 5...

from autograd import grad
grad_cost= grad(cost)

grad_cost(W, H)
# The above returns an array of shape of W
```

**Comment by j-towns** [comment]
Checkout the [convenience_wrappers](https://github.com/HIPS/autograd/blob/master/autograd/convenience_wrappers.py). Either `multigrad` or `multigrad_dict` should do the trick.

**Comment by nipunbatra** [comment]
Thanks a lot for the quick reply. Does the following usage look correct to you? It does give me the expected answer. If it seems correct, I'm happy to close the issue and happy to do a PR if you'd like this included in the example set.

```python
import autograd.numpy as np
import pandas as pd
import matplotlib.pyplot as plt

A = np.array([[3, 4, 5, 2],
                   [4, 4, 3, 3],
                   [5, 5, 4, 4]], dtype=np.float32).T

shape = A.shape
rank = 2

def cost(W, H):
    pred = np.dot(W, H)
    return np.sqrt(((pred - A).flatten() ** 2).mean(axis=None))

from autograd import grad, multigrad

grad_cost= multigrad(cost, argnums=[0,1])

H =  np.random.randn(rank, shape[1])
W =  np.random.randn(shape[0], rank)

learning_rate=0.01

for i in range(5000):
    if i%50==0:
        print "*"*20
        print i
        print "*"*20
        print cost(W, H)
    del_W, del_H = grad_cost(W, H)
    W = W-del_W*learning_rate
    H = H-del_H*learning_rate

pred = np.dot(W, H)
```

**Comment by MInner** [comment]
There's also [`autograd.grad_named`](https://github.com/HIPS/autograd/blob/master/autograd/convenience_wrappers.py#L67) that works pretty much as you expect: 

`d_cost_dW = autograd.grad_named(cost, 'W')`
